typing/typecore.ml

```diff
@@ -6345,6 +6344,63 @@ let type_clash_of_trace trace =
     | _ -> None
   ))
 
+(** More precise denomination for type errors. Used by messages:
+
+    - "This <denom> ..."
+    - "The <denom> 'foo' ..." *)
+let pp_exp_denom ppf exp =
+  let f = fprintf ppf "%s" in
+  let desc_expression = fprintf ppf "%a expression" Style.inline_code in
+  match exp.pexp_desc with
+  | Pexp_constant _ -> f "constant"
+  | Pexp_ident _ -> f "value"
+  | Pexp_construct _ | Pexp_variant _ -> f "constructor"
+  | Pexp_field _ -> f "field access"
+  | Pexp_send _ -> f "method call"
+  | Pexp_while _ -> desc_expression "while"
+  | Pexp_for _ -> desc_expression "for"
+  | Pexp_ifthenelse _ -> desc_expression "if-then-else"
+  | Pexp_match _ -> desc_expression "match"
+  | Pexp_try _ -> desc_expression "try-with"
+  | _ -> f "expression"
+
+let report_general_this_exp denom ppf = fprintf ppf "This %t" denom
+
+let pp_inline_exp =
+  let pp_exp ppf exp =
+    match exp.pexp_desc with
+    | Pexp_constant { pconst_desc = Pconst_string (s, _, None); _ } ->
+        (* Avoid adding quotes around a quoted string constant but make sure to
+           use consistent inline code style. *)
+        Format.pp_print_string ppf (String.escaped s)
+    | _ -> Pprintast.expression ppf exp
+  in
+  Style.as_inline_code pp_exp
+
+(** Implements the "This expression" message, printing the expression if it
+    should be according to {!Parsetree.exp_is_nominal}. *)
+let report_this_exp denom ppf exp =
+  let denom ppf =
+    match denom with
+    | Some d -> fprintf ppf "%s" d
+    | None -> pp_exp_denom ppf exp
+  in
+  if Pprintast.exp_is_nominal exp then
+    fprintf ppf "The %t %a" denom pp_inline_exp exp
+  else report_general_this_exp denom ppf
+
+let report_this_exp_opt denom ppf = function
+  | Some exp -> report_this_exp denom ppf exp
+  | None ->
+      let denom ppf =
+        let d = match denom with Some d -> d | None -> "expression" in
+        fprintf ppf "%s" d
+      in
+      report_general_this_exp denom ppf
```

Comment:
It seems to me that the two `report_this_exp` function could be merged without loss of readability.

Reply:
Do you mean change `report_this_pexp` to take an option `pexp` ?

Reply:
Yes, I was thinking of making the choice of denomination in one place:
```ocaml
let report_this_pexp_opt denom ppf exp =
  let denom ppf =
    match denom, exp with
    | Some d, _ -> fprintf ppf "%s" d
    | None, Some exp -> pp_pexp_denom ppf exp
    | None, None -> fprintf ppf "expression"
  in
  let nexp = Option.bind exp Pprintast.Doc.(nominal_exp_doc longident) in
  match nexp with
  | Some nexp -> fprintf ppf "The %t %a" denom pp_doc nexp
  | _ -> fprintf ppf "This %t" denom
  ```

Reply:
Looking at the call sites, we could also make the sentence more readable by adding the `has type` suffix.

Reply:
You removed the wrapping of `nexp` with `Style.as_inline_code` in your snippet. Is that intended ?

Reply:
No, that was a mistake.

Reply:
Thanks, I've changed the code.